pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ThresholdLoadManagerStrategy.java

```diff
@@ -0,0 +1,131 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.loadbalance.impl;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.broker.ServiceConfiguration;
+import org.apache.pulsar.broker.loadbalance.LoadData;
+import org.apache.pulsar.broker.loadbalance.ModularLoadManagerStrategy;
+import org.apache.pulsar.policies.data.loadbalancer.BrokerData;
+import org.apache.pulsar.policies.data.loadbalancer.BundleData;
+import org.apache.pulsar.policies.data.loadbalancer.LocalBrokerData;
+
+/**
+ * Placement strategy which randomly selects a broker from those having resource usage below average load.
+ */
+@Slf4j
+public class ThresholdLoadManagerStrategy implements ModularLoadManagerStrategy {
+    // Maintain this list to reduce object creation.
+    private final ArrayList<String> bestBrokers;
+
+    // candidate broker -> broker usage.
+    private final Map<String, Double> brokerUsages;
+
+    public ThresholdLoadManagerStrategy() {
+        bestBrokers = new ArrayList<>();
+        brokerUsages = new HashMap<>();
+    }
+
+    private static double getScore(final BrokerData brokerData, final ServiceConfiguration conf) {
+        final double overloadThreshold = conf.getLoadBalancerBrokerOverloadedThresholdPercentage() / 100.0;
+        final double maxUsage = brokerData.getLocalData().getMaxResourceUsage();
+        if (maxUsage > overloadThreshold) {
+            log.warn("Broker {} is overloaded: max usage={}", brokerData.getLocalData().getWebServiceUrl(), maxUsage);
+            return Double.POSITIVE_INFINITY;
+        }
+
+        double brokerResourceUsage = brokerData.getLocalData().getMaxResourceUsageWithWeight(
```

Comment:
Afaik, this `brokerResourceUsage` is relatively short-term data(every 5 sec).  Can you share why we are using this short-term usage data in this `getScore()` function?

For example, I am wondering how we are going to handle the following scenarios. 
- In one sample, one broker was temporally idle, but afterward, it became mostly busy.
- In one sample, conversely, one broker was temporally very busy, but afterward, it became mostly idle.

I don't think there is a definitive answer here, but could you share why we only use the short-term usage data here?

Reply:
You are right. `brokerResourceUsage` is relatively short-term data.

As for your question, using short-term data is also ok. Because the strategy here only affect scenarios where new traffic needs to be loaded. Once the broker is busy, it should not be selected for new traffic. Once the broker is idle, it is good to be selected for new traffic as well. Once more, there is random selection logic here to prevent a lot of traffic from being loaded on a single broker.

Reply:
- I am wondering if the short-term data is sufficient to check the load of the brokers. (here we skip brokers whose short-term usage is bigger than the threshold). 

- Another question is, why do we return `getMaxResourceUsageWithWeight` although we check the threshold with `getMaxResourceUsage`?


- I think the random selection is fine.

Reply:
- It is a open question. I think using short-term data is ok, because the strategy here will choose the broker which is under average load just at the time that pulsar need to load bundle. Using long-term data might be worse that short-term data if there are lots of traffic fluctuations in a short time.
- We check the threshold with `getMaxResourceUsage` to check whether the broker is overloaded, and return `getMaxResourceUsageWithWeight` as the broker load score just to align the implementation of `org.apache.pulsar.broker.loadbalance.impl.ThresholdShedder`.

Reply:
- I agree that this is an open question. It's up to the users' choice in the end. Since other load balance logics use historical usage, I think we can point it out here. For example, ThresholdShedder uses historical load `e.g. usage = prev_usage * 0.9 + (1.0) * current_usage. `
- Afaik, ThresholdShedder only uses `getMaxResourceUsageWithWeight`. Again, why don't we check the threshold with `getMaxResourceUsageWithWeight`?

Reply:
> * Afaik, ThresholdShedder only uses `getMaxResourceUsageWithWeight`. Again, why don't we check the threshold with `getMaxResourceUsageWithWeight`?

I realized that `LeastLongTermMessageRate` uses `getMaxResourceUsage`. I think that's what you meant before.